File: terminal_ticker/domain/price_action.py
"""文件用途：领域层，定义标准化 K 线数据。"""
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class Candle:
    """说明：封装一根标准化 OHLCV K 线。"""
    symbol_key: str
    open_time_ms: int
    open: float
    high: float
    low: float
    close: float
    volume: float

    def __post_init__(self) -> None:
        """说明：在数据对象创建后校验字段合法性。"""
        if self.high < max(self.open, self.close, self.low):
            raise ValueError("candle high must be greater than or equal to open, low, and close")
        if self.low > min(self.open, self.close, self.high):
            raise ValueError("candle low must be less than or equal to open, high, and close")

    @property
    def range(self) -> float:
        """说明：返回 K 线最高价和最低价之间的范围。"""
        return self.high - self.low

    @property
    def body(self) -> float:
        """说明：返回 K 线实体大小。"""
        return abs(self.close - self.open)

File: tradex/domain/price_action.ts
export interface CandleInput {
  symbolKey: string;
  openTimeMs: number;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

export class Candle {
  readonly symbolKey: string;
  readonly openTimeMs: number;
  readonly open: number;
  readonly high: number;
  readonly low: number;
  readonly close: number;
  readonly volume: number;

  constructor(input: CandleInput) {
    if (input.high < Math.max(input.open, input.close, input.low)) {
      throw new Error("candle high must be greater than or equal to open, low, and close");
    }
    if (input.low > Math.min(input.open, input.close, input.high)) {
      throw new Error("candle low must be less than or equal to open, high, and close");
    }
    this.symbolKey = input.symbolKey;
    this.openTimeMs = input.openTimeMs;
    this.open = input.open;
    this.high = input.high;
    this.low = input.low;
    this.close = input.close;
    this.volume = input.volume;
  }

  get range(): number {
    return this.high - this.low;
  }

  get body(): number {
    return Math.abs(this.close - this.open);
  }
}

export function mergeCandles(existing: readonly Candle[], incoming: readonly Candle[]): Candle[] {
  const byOpenTime = new Map<number, Candle>();
  for (const candle of existing) byOpenTime.set(candle.openTimeMs, candle);
  for (const candle of incoming) byOpenTime.set(candle.openTimeMs, candle);
  return [...byOpenTime.entries()]
    .sort(([left], [right]) => left - right)
    .map(([, candle]) => candle);
}
